Python 3 support added without losing Python 2.7

moigagoo opened the pull request:
First off, thanks for the fantastic extension!

I've added Python 3 support to it.

moigagoo commented:
Important: Autodoc doesn't work yet.

c2nes commented on javasphinx/compiler.py:
```diff
@@ -255,7 +255,7 @@ def compile_type_document(self, imports_block, package, name, declaration):
         if fields:
             document.add_heading('Fields', '-')
-            fields.sort(key=lambda f: f.declarators[0].name)
+            list(fields).sort(key=lambda f: f.declarators[0].name)
```
Unfortunately this won't sort the contents of `fields`. `list(fields).sort(...)` creates and then sorts a _copy_ of the list, but leaves the original iterator returned by `filter` unmodified. I think we should instead wrap the `filter` call on line 255 with `list(...)`. This will also fixes an issue with the conditional on line 256. In 2.x `filter` returns a list, so the conditional will fail if the list is empty. In 3.x `filter` returns an iterator which always evaluates as true.

c2nes commented on javasphinx/compiler.py:
```diff
@@ -266,7 +266,7 @@ def compile_type_document(self, imports_block, package, name, declaration):
         if constructors:
             document.add_heading('Constructors', '-')
-            constructors.sort(key=lambda c: c.name)
+            list(constructors).sort(key=lambda c: c.name)
```
Same here

c2nes commented on javasphinx/compiler.py:
```diff
@@ -277,7 +277,7 @@ def compile_type_document(self, imports_block, package, name, declaration):
         if methods:
             document.add_heading('Methods', '-')
-            methods.sort(key=lambda m: m.name)
+            list(methods).sort(key=lambda m: m.name)
```
And here as well

c2nes commented on javasphinx/htmlrst.py:
```diff
@@ -233,7 +235,7 @@ def _process_text(self, node):
 
     def _process(self, node):
-        if isinstance(node, basestring):
+        if isinstance(node, str):
```
I think this change may cause problems with unicode strings in 2.x. Using the support from [python-future](http://python-future.org) for str/unicode handling throughout might be a good idea.

c2nes commented:
Hey miogagoo, glad you like the extension and thanks for adding Python 3 support!

Overall this looks pretty good. I left a few comments concerning `list.sort` and `str` vs `unicode`. I think the str/unicode stuff would benefit from using one of the 2/3 compatibility libraries. I've been looking at [python-future](http://python-future.org) which seems pretty nice.

moigagoo commented:
@c2nes Thanks a lot for the review! My bad, I was in a hurry and didn't test my modifications as thoroughly as I should have. I will go through your comments and fix the code as soon as I can.

moigagoo commented:
@c2nes I've addressed the issues you had pointed out.

c2nes commented on javasphinx/util.py:
```diff
@@ -14,13 +14,15 @@
 #
 
+from __future__ import unicode_literals
+
```
Lets add `from builtins import str` here so this is consistently unicode between Python 2.x and 3.x

c2nes commented on javasphinx/htmlrst.py:
```diff
@@ -395,8 +397,8 @@ def _preprocess(self, s_html):
-        if not isinstance(s_html, unicode):
-            s_html = unicode(s_html, 'utf8')
+        if not isinstance(s_html, str):
+            s_html = str(s_html, 'utf8')
```
The meaning of `str` here will vary between Python 2.x and 3.x. I think we want `from builtins import str` at the top of this file and this will be good. I'd also recommend removing the `basestring` import and updating the `_process` method to check `isinstance(node, str)` instead of `isinstance(node, basestring)`.

c2nes commented on javasphinx/htmlrst.py:
```diff
@@ -14,6 +14,8 @@
 # limitations under the License.
 #
 
+from past.builtins import basestring
```
`from __future__ import unicode_literals` would also be good here

c2nes commented:
Sorry for not reviewing sooner. Couple more small comments, but I think those should do it.

moigagoo commented:
@c2nes Thanks for the review! I've addressed the issues you had outlines in this commit: https://github.com/bronto/javasphinx/pull/39/commits/be6056e76e001bfd83a727e89d48763a229da8af

c2nes commented:
Looks good @moigagoo. Thanks again for the contribution!

roipoussiere commented:
Sorry, Python3 support do not work for me. :/ #51
